pallets/time-release/src/lib.rs

```diff
@@ -0,0 +1,449 @@
+
+			Self::deposit_event(Event::Claimed { who, amount: locked_amount });
+			Ok(())
+		}
+
+		/// Add a new release schedule for an account.
+		///
+		/// # Events
+		/// * [`Event::ReleaseScheduleAdded]
+		///
+		/// # Errors
+		///
+		/// * [`Error::InsufficientBalanceToLock] - Insuffience amount of balance to lock.
+		/// * [`Error::MaxReleaseSchedulesExceeded] - Failed because the maximum release schedules was exceeded-
+		/// * [`ArithmeticError::Overflow] - Failed because of an overflow.
+		///
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::transfer())]
+		pub fn transfer(
+			origin: OriginFor<T>,
+			dest: <T::Lookup as StaticLookup>::Source,
+			schedule: ReleaseScheduleOf<T>,
+		) -> DispatchResult {
+			let from = T::TransferOrigin::ensure_origin(origin)?;
+			let to = T::Lookup::lookup(dest)?;
+
+			if to == from {
+				ensure!(
+					T::Currency::free_balance(&from) >=
+						schedule.total_amount().ok_or(ArithmeticError::Overflow)?,
+					Error::<T>::InsufficientBalanceToLock,
+				);
+			}
+
+			Self::do_transfer(&from, &to, schedule.clone())?;
+
+			Self::deposit_event(Event::ReleaseScheduleAdded {
+				from,
+				to,
+				release_schedule: schedule,
+			});
+			Ok(())
+		}
+
+		/// Update all release schedules under an account, `root` origin required.
+		///
+		/// # Events
+		/// * [`Event::ReleaseSchedulesUpdated] - Updated release schedules.
+		///
+		/// # Errors
+		///
+		/// * [`Error::InsufficientBalanceToLock] - Insuffience amount of balance to lock.
+		/// * [`Error::MaxReleaseSchedulesExceeded] - Failed because the maximum release schedules was exceeded-
+		/// * [`ArithmeticError::Overflow] - Failed because of an overflow.
+		///
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::update_release_schedules(release_schedules.len() as u32))]
+		pub fn update_release_schedules(
+			origin: OriginFor<T>,
+			who: <T::Lookup as StaticLookup>::Source,
+			release_schedules: Vec<ReleaseScheduleOf<T>>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let account = T::Lookup::lookup(who)?;
+			Self::do_update_release_schedules(&account, release_schedules)?;
+
+			Self::deposit_event(Event::ReleaseSchedulesUpdated { who: account });
+			Ok(())
+		}
+
+		/// Claim unlocked balances on behalf for an account.
+		///
+		/// # Events
+		/// * [`Event::Claimed`]
+		///
+		#[pallet::call_index(3)]
+		#[pallet::weight(T::WeightInfo::claim(<T as Config>::MaxReleaseSchedules::get() / 2))]
+		pub fn claim_for(
+			origin: OriginFor<T>,
+			dest: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			let who = T::Lookup::lookup(dest)?;
+			let locked_amount = Self::do_claim(&who);
+
+			Self::deposit_event(Event::Claimed { who, amount: locked_amount });
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn do_claim(who: &T::AccountId) -> BalanceOf<T> {
+		let locked = Self::locked_balance(who);
+		if locked.is_zero() {
+			Self::clear_release_schedules(who);
```

Comment:
it seems like we are removing the same storage twice once in `locked_balance` and once in `clear_release_schedules`

Reply:
If not counted in the benchmark it might be fine.

Reply:
Yeah, but it is confusing.

Reply:
I made some updates. Check it out.